### PR TITLE
YQL-18104: add support for yql resources in arrow blocks

### PR DESCRIPTION
**Pull request by fedor-miron**
### Changelog entry 

Added support for YQL resources in Arrow format. Added managed Arrow buffer for types with non-trivial destructors

### Changelog category 

* Improvement

### CODE REVIEW

File: ydb/library/yql/minikql/mkql_type_builder.cpp [hunk 1]
```diff
@@ -2437,6 +2446,7 @@ struct TComparatorTraits {
     template <typename TStringType, bool Nullable, NUdf::EDataSlot TOriginal = NUdf::EDataSlot::String>
     using TStrings = NUdf::TStringBlockItemComparator<TStringType, Nullable>;
     using TExtOptional = NUdf::TExternalOptionalBlockItemComparator;
+    static const bool ImplementedForResources = false;
```
Comment (vitstn) [inline ydb/library/yql/minikql/mkql_type_builder.cpp:2449]: Unclear semantic of such flag

File: ydb/library/yql/public/udf/arrow/block_builder.h [hunk 2]
```diff
@@ -466,10 +486,14 @@ class TFixedSizeArrayBuilder final : public TArrayBuilderBase {
         }
     }
 
+    void DoAdd(TLayout&& value)  {
+        ::new(DataPtr + GetCurrLen()) TLayout(std::move(value));
+    }
+
     std::unique_ptr<TTypedBufferBuilder<ui8>> NullBuilder;
-    std::unique_ptr<TTypedBufferBuilder<T>> DataBuilder;
+    std::unique_ptr<TTypedBufferBuilder<TLayout>> DataBuilder;
```
Comment (vitstn) [inline ydb/library/yql/public/udf/arrow/block_builder.h:494]: destructor should call destructor of UV in DataBuilder

File: ydb/library/yql/public/udf/arrow/block_builder.h [hunk 3]
```diff
@@ -309,7 +309,21 @@ class TArrayBuilderBase : public IArrayBuilder {
...
+template<typename T>
+T FromBlockItemDefault(TBlockItem item) {
+    return item.Get<T>(); 
+};
+}
+
+template<typename TLayout, bool Nullable, 
+TLayout FromUnboxedValuePod(TUnboxedValuePod) = FromUnboxedValuePodDefault, 
+TLayout FromBlockItem(TBlockItem) = FromBlockItemDefault>
 class TFixedSizeArrayBuilder final : public TArrayBuilderBase {
```
Comment (nepal) [inline ydb/library/yql/public/udf/arrow/block_builder.h:327]: Мне кажется надо сделать отдельный билдер для ресурсов. Как минимум новый билдер должен деструкторы звать

**Review by nepal — DISMISSED** [review]

File: ydb/library/yql/public/udf/arrow/block_builder.h [hunk 4]
```diff
@@ -1091,6 +1115,25 @@ std::unique_ptr<TArrayBuilderBase> MakeArrayBuilderBase(
     const ITypeInfoHelper& typeInfoHelper, const TType* type, arrow::MemoryPool& pool, 
     size_t maxBlockLength, const IPgBuilder* pgBuilder);
 
+template<bool Nullable>
+std::unique_ptr<TArrayBuilderBase> MakeResourceArrayBuilder(
+    const ITypeInfoHelper& typeInfoHelper, const TType* type, arrow::MemoryPool& pool, 
+    size_t maxLen) {
+
+    using TBuilder = TFixedSizeArrayBuilder<TUnboxedValue, Nullable,
```
Comment (vitstn) [inline ydb/library/yql/public/udf/arrow/block_builder.h:1123]: please use distinct class TResourceArrayBuilder - it has too many differences to POD builder

File: ydb/library/yql/public/udf/arrow/block_item.h [hunk 5]
```diff
@@ -64,25 +67,30 @@ class TBlockItem {
...
         return Raw.Tuple.Value[index];
     }
 
     inline TStringRef AsStringRef() const {
-        Y_DEBUG_ABORT_UNLESS(Raw.GetMarkers() == EMarkers::Present);
+        Y_DEBUG_ABORT_UNLESS(GetMarkers() == EMarkers::String);
         return TStringRef(Raw.String.Value, Raw.String.Size);
     }
 
+    inline TStringRef GetEmbeddedStringRef() const {
```
Comment (vitstn) [inline ydb/library/yql/public/udf/arrow/block_item.h:89]: I don't see any constructor that can initialize such state: EMarkers::Embedded. BlockItem(TStringRef value) sets EMarkers::String

File: ydb/library/yql/public/udf/arrow/block_builder.h (hunk 3 above)
Comment (fedor-miron) [reply ydb/library/yql/public/udf/arrow/block_builder.h:327]: Деструкторы зовет ManagedBuffer здесь https://github.com/ydb-platform/ydb/pull/3116/files#diff-e636d2a5068525a81197ad5b1595c3716f92f9109aada6a0e27a834b9694da54R114
Билдеру не нужно звать деструкторы

**Comment by fedor-miron** [comment]
Все поправил

**Review by vitstn — DISMISSED** [review]

File: ydb/library/yql/minikql/computation/mkql_block_reader.cpp [hunk 6]
```diff
@@ -39,6 +39,43 @@ class TFixedSizeBlockItemConverter : public IBlockItemConverter {
...
+public:
+    NUdf::TUnboxedValuePod MakeValue(TBlockItem item, const THolderFactory& holderFactory) const final {
+        Y_UNUSED(holderFactory);
+        if constexpr (Nullable) {
+            if (!item) {
+                return {};
+            }
+        }
+
+        if (item.IsEmbedded()) {
```
Comment (vitstn) [inline ydb/library/yql/minikql/computation/mkql_block_reader.cpp:53]: resource can be a StringValue too

File: ydb/library/yql/minikql/computation/mkql_block_reader.cpp [hunk 7]
```diff
@@ -39,6 +39,43 @@ class TFixedSizeBlockItemConverter : public IBlockItemConverter {
...
+    }
+
+    TBlockItem MakeItem(const NUdf::TUnboxedValuePod& value) const final {
+        if constexpr (Nullable) {
+            if (!value) {
+                return {};
+            }
+        }
+
+        if (value.IsEmbedded()) {
```
Comment (vitstn) [inline ydb/library/yql/minikql/computation/mkql_block_reader.cpp:69]: the same

**Review by nepal — DISMISSED** [review]
LGTM

**Review by vitstn — APPROVED** [review]

**Pull request merged**